alibi_detect/od/_svm.py

```diff
@@ -0,0 +1,233 @@
+
+import numpy as np
+
+from alibi_detect.base import (BaseDetector, FitMixin, ThresholdMixin,
+                               outlier_prediction_dict)
+from alibi_detect.exceptions import _catch_error as catch_error
+from alibi_detect.od.pytorch import DgSVMTorch, SdgSVMTorch
+from alibi_detect.utils._types import Literal
+from alibi_detect.utils.frameworks import BackendValidator
+from alibi_detect.version import __version__
+
+if TYPE_CHECKING:
+    import torch
+
+
+backends = {
+    'pytorch': {
+        'sgd': SdgSVMTorch,
+        'gd': DgSVMTorch
+    }
+}
+
+
+class SVM(BaseDetector, ThresholdMixin, FitMixin):
+    def __init__(
+        self,
+        nu: float,
+        kernel: 'torch.nn.Module',
+        n_components: Optional[int],
+        optimization: Literal['sgd', 'gd'] = 'sgd',
+        backend: Literal['pytorch'] = 'pytorch',
+        device: Optional[Union[Literal['cuda', 'gpu', 'cpu'], 'torch.device']] = None,
+    ) -> None:
+        """One-Class Support vector machine (OCSVM) outlier detector.
+
+        The one-class Support vector machine outlier detector fits a one-class SVM to the reference data.
+
+        Rather than the typical approach of optimizing the exact kernel OCSVM objective through a dual formulation,
+        here we instead map the data into the kernel's RKHS and then solve the linear optimization problem
+        directly through its primal formulation. The Nystroem approximation is used to speed up training and inference
+        by approximating the kernel's RKHS.
+
+        We provide two options, specified by the `optimization` parameter, for optimizing the one-class svm. `''sgd''`
+        wraps the `SGDOneClassSVM` class from the sklearn package and the other, `''gd''` uses a custom implementation
+        in PyTorch. The PyTorch approach is tailored for operation on GPUs. Instead of applying stochastic gradient
+        descent (one data point at a time) with a fixed learning rate schedule it performs full gradient descent with
+        step size chosen at each iteration via line search. Note that on a CPU this would not necessarily be preferable
+        to SGD as we would have to iterate through both data points and candidate step sizes, however on GPU all of the
+        operations are vectorized/parallelized. Moreover, the Nystroem approximation has complexity `O(n^2m)` where
+        `n` is the number of reference instances and `m` defines the number of inducing points. This can therefore be
+        expensive for large reference sets and benefits from implementation on the GPU.
+
+        In general if using a small dataset then using the `''cpu''` with the optimization `''sgd''` is the best choice.
+        Whereas if using a large dataset then using the `''gpu''` with the optimization `''gd''` is the best choice.
+
+        Parameters
+        ----------
+        nu
+            The proportion of the training data that should be considered outliers. Note that this does not necessarily
```

Comment:
Lets make clear this should be thought of as a regularisation parameter that affects how smooth the decsion boundary will be.

Reply:
Ok, I've basically just added your comment on at the end:

> The proportion of the training data that should be considered outliers. Note that this does not necessarily correspond to the false positive rate on test data, which is still defined when calling the `infer_threshold` method. `nu` should be thought of as a regularization parameter that affects how smooth the svm decision boundary is.